File: jlox/src/lox/Lox.java
package lox;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Lox {
    private static final Interpreter interpreter = new Interpreter();
    private static boolean hadError = false;
    private static boolean hadRuntimeError = false;
    private static boolean debug = false;

    public static void main(String[] args) throws IOException {
        switch (args.length) {
            case 0 -> {
                debug = false;
                runPrompt();
            }

            case 1 -> {
                if (args[0].equals("-d")) {
                    debug = true;
                    runPrompt();
                } else {
                    debug = false;
                    runFile(args[0]);
                }
            }

            case 2 -> {
                if (args[0].equals("-d")) {
                    debug = true;
                    runFile(args[1]);
                } else {
                    usage();
                }
            }

            default -> usage();
        }
    }

    private static void runFile(String path) throws IOException {
        byte[] bytes = Files.readAllBytes(Paths.get(path));
        run(new String(bytes, Charset.defaultCharset()));

        if (hadError)
            System.exit(65);
        if (hadRuntimeError)
            System.exit(70);
    }

    private static void runPrompt() throws IOException {
        InputStreamReader input = new InputStreamReader(System.in);
        BufferedReader reader = new BufferedReader(input);

        for (;;) {
            System.out.print(">>> ");
            String line = reader.readLine();

            if (line == null) {
                System.out.println();
                break;
            }

            run(line);

            hadError = false;
            hadRuntimeError = false;
        }
    }

    private static void run(String source) {
        Scanner scanner = new Scanner(source);
        List<Token> tokens = scanner.scanTokens();

        Parser parser = new Parser(tokens);
        List<Stmt> statements = parser.parse();

        if (hadError)
            return;

        if (debug) {
            // Tokenization
            for (Token token : tokens) {
                System.out.println(token);
            }

            // Ast Representation
            for (Stmt statement : statements) {
                System.out.println(new AstPrinter().print(statement));
            }
        } else {
            Resolver resolver = new Resolver(interpreter);
            resolver.resolve(statements);

            if (hadError)
                return;

            interpreter.interpret(statements);
        }
    }

    static void usage() {
        System.err.println("Usage: jlox *-d <script>");
        System.err.println("* '-d' is an optional flag for enabling debug info.");
        System.err.println("  This will suppress interpretation, dumping a print-out of");
        System.err.println("  the lexical token generation and the AST-representation generated");
        System.err.println("  by the Parser.");
        System.exit(64);
    }

    static void error(int line, String message) {
        report(line, "", message);
    }

    private static void report(int line, String where, String message) {
        System.err.println("[line " + line + "] Error" + where + ": " + message);
        hadError = true;
    }

    static void error(Token token, String message) {
        if (token.type == TokenType.EOF) {
            report(token.line, " at end", message);
        } else {
            report(token.line, " at '" + token.lexeme + "'", message);
        }
    }

    static void runtimeError(RuntimeError error) {
        System.err.println(error.getMessage() +
                "\n[line " + error.token.line + "]");
        hadRuntimeError = true;
    }
}

File: plox/lox.py
#!/usr/bin/env python3

# Python Imports
import sys

# Project Imports
from src.ast.printer import AstPrinter
from src.interpreter.interpreter import Interpreter
from src.parser.parser import Parser
from src.parser.resolver import Resolver
from src.scanner.scanner import Scanner
from src.util.errors import ErrType, LoxError
from src.util.mode import RunMode


class Lox:
    def __init__(self):
        self.file_name = "STDIN"
        self.debug = False
        self.err_manager = LoxError()
        self.mode = RunMode.FILE
        self.interpreter = Interpreter(self.err_manager, self.mode)

    def main(self):
        exec, *argv = sys.argv

        match len(argv):
            case 0:
                self.mode = RunMode.REPL
                self.run_prompt()
            case 1:
                arg, *argv = argv

                if arg == "-d":
                    self.debug = True
                    self.mode = RunMode.REPL

                    self.run_prompt()
                else:
                    self.debug = False
                    self.run_file(arg)
            case 2:
                cmd, *argv = argv
                fp, *argv = argv

                if cmd == "-d":
                    self.debug = True
                    self.run_file(fp)
                else:
                    self.err_manager.error(
                        ErrType.ARG_ERROR,
                        f"Invalid command provided: {cmd}"
                    )
                    self.usage()
                    sys.exit(64)
            case _:
                self.err_manager.error(
                    ErrType.ARG_ERROR,
                    "Too many arguments provided."
                )
                self.usage()
                sys.exit(64)

    def run_file(self, file_path):
        try:
            with open(file_path, "r") as f:
                source = f.read()

            self.run(source)

            if self.err_manager.had_error:
                sys.exit(65)
            if self.err_manager.had_runtime_error:
                sys.exit(70)

        except FileNotFoundError as e:
            print(e)
            self.err_manager.error(ErrType.IO_ERROR, f" {e}"[10:])

    def run_prompt(self):
        if self.debug:
            print("Lox REPL Version 0.0.1 [DEBUG MODE]")
            print("Lang Version 0.0.5")
        else:
            print("Lox REPL Version 0.0.1")
            print("Lang Version 0.0.5")
        print("Press Ctrl-D to quit.")

        try:
            while True:
                line = input(">>> ")

                if line == "exit":
                    break

                self.run(line)

                self.err_manager.had_error = False
                self.err_manager.had_runtime_error = False

        except EOFError:
            print()
            sys.exit(0)

        except KeyboardInterrupt:
            print()
            sys.exit(0)

    def run(self, source):
        # Lexer
        scanner = Scanner(source, self.err_manager)
        tokens = scanner.scan_tokens()

        if self.err_manager.had_error:
            return

        # Parser
        parser = Parser(tokens, self.err_manager)
        statements = parser.parse()

        if self.err_manager.had_error:
            if self.debug:
                for token in tokens:
                    print(token, file=sys.stderr)
            return

        if self.debug:
            printer = AstPrinter()
            for token in tokens:
                print(token, file=sys.stderr)

            for statement in statements:
                printer.print_stmt(statement)
        else:
            resolver = Resolver(self.interpreter, self.err_manager)
            resolver.resolve_stmts(statements)

            if self.err_manager.had_error:
                return

            self.interpreter.mode = self.mode
            self.interpreter.interpret(statements)

    def usage(self):
        print("Usage: ./plox.py *-d <script>")
        print("* Optional command for enabling debugging info for your script.")
        print("  This suppresses the interpreter and prints out the lexical breakdown")
        print("  as well as the AST representation generated by the parser.")


if __name__ == "__main__":
    lox = Lox()
    lox.main()
